Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs

```diff
@@ -26,8 +25,37 @@ public abstract class SMTLibProcessTheoremProver : ProverInterface
     protected TypeAxiomBuilder CachedAxBuilder;
     internal abstract ScopedNamer Namer { get; }
     protected TypeDeclCollector DeclCollector;
-    protected readonly List<string> proverErrors = new();
-    protected readonly List<string> proverWarnings = new();
+
+    private readonly ConditionalWeakTable<SMTLibSolver, ConcurrentStack<string>> allProverErrors = new();
```

Comment:
Alternatively, you could change move `HandleProverError` into a new class `ProverProblemCollector` that contains the errors and warnings, and create a new collector each time you create a new process:

```csharp
    protected void SetupProcess()
    {
      Process?.Close();
      Process = options.Solver == SolverKind.NoOpWithZ3Options
        ? new NoopSolver()
        : new SMTLibProcess(libOptions, options);
      this.proverProblemCollector = new ProverProblemCollector(this);
      Process.ErrorHandler += proverProblemCollector.HandleProverError;
    }
```

Reply:
I like this better. I'm not fan of conditional weak tables, so this looks like a good solution.